### PR TITLE
[Arc] Add LowerVectorizations pass

### PR DESCRIPTION
**Pull request by maerhart**
Adds a pass to lower  operations. Currently, only lowering the boundary and inlining the op body is supported. Lowering of the body will be added in a separate commit since that is a bit more complex.

### CODE REVIEW

File: include/circt/Dialect/Arc/ArcPasses.td [hunk 1]
```diff
@@ -161,6 +162,79 @@ def LowerState : Pass<"arc-lower-state", "mlir::ModuleOp"> {
...
+    be fully lowered (i.e., the op disappears in the IR). Alternatively, it can
+    be partially lowered.
+    
+    The "mode" pass option allows to only lower the boundary, only the body, or
+    only inline the body given that both the boundary and the body are already
+    lowered.
+
+    The pass supports vectorization within scalar registers and SIMD
+    vectorization and prioritizes vectorization by packing the vector elements
+    into a scalar value if it can fit into 64 bits.
```
Comment (fabianschuiki) [inline include/circt/Dialect/Arc/ArcPasses.td:188]: What a beast of a pass :sunglasses:

File: lib/Dialect/Arc/Transforms/AddTaps.cpp [hunk 2]
```diff
@@ -6,15 +6,24 @@
...
+} // namespace arc
+} // namespace circt
 
 using namespace circt;
 using namespace arc;
 using namespace hw;
 
 namespace {
-struct AddTapsPass : public AddTapsBase<AddTapsPass> {
+struct AddTapsPass : public arc::impl::AddTapsBase<AddTapsPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/AddTaps.cpp:26]: Unrelated changes

File: lib/Dialect/Arc/Transforms/AllocateState.cpp [hunk 3]
```diff
@@ -22,7 +32,8 @@ using llvm::SmallMapVector;
 //===----------------------------------------------------------------------===//
 
 namespace {
-struct AllocateStatePass : public AllocateStateBase<AllocateStatePass> {
+struct AllocateStatePass
+    : public arc::impl::AllocateStateBase<AllocateStatePass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/AllocateState.cpp:36]: Unrelated changes

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp [hunk 4]
```diff
@@ -486,7 +494,7 @@ SinkArcInputsPattern::matchAndRewrite(DefineOp op,
 
 namespace {
 struct ArcCanonicalizerPass
-    : public ArcCanonicalizerBase<ArcCanonicalizerPass> {
+    : public arc::impl::ArcCanonicalizerBase<ArcCanonicalizerPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:497]: Unrelated changes

File: lib/Dialect/Arc/Transforms/Dedup.cpp [hunk 5]
```diff
@@ -351,7 +360,7 @@ static bool isOutlinable(OpOperand &operand) {
 }
 
 namespace {
-struct DedupPass : public DedupBase<DedupPass> {
+struct DedupPass : public arc::impl::DedupBase<DedupPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/Dedup.cpp:363]: Unrelated changes

File: lib/Dialect/Arc/Transforms/GroupResetsAndEnables.cpp [hunk 6]
```diff
@@ -182,7 +190,7 @@ struct GroupAssignmentsInIfPattern : public OpRewritePattern<scf::IfOp> {
 
 namespace {
 struct GroupResetsAndEnablesPass
-    : public GroupResetsAndEnablesBase<GroupResetsAndEnablesPass> {
+    : public arc::impl::GroupResetsAndEnablesBase<GroupResetsAndEnablesPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/GroupResetsAndEnables.cpp:193]: Unrelated changes

File: lib/Dialect/Arc/Transforms/InferMemories.cpp [hunk 7]
```diff
@@ -6,21 +6,32 @@
...
+} // namespace arc
+} // namespace circt
+
 using namespace circt;
 using namespace arc;
 
 namespace {
-struct InferMemoriesPass : public InferMemoriesBase<InferMemoriesPass> {
+struct InferMemoriesPass
+    : public arc::impl::InferMemoriesBase<InferMemoriesPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/InferMemories.cpp:34]: Unrelated changes

File: lib/Dialect/Arc/Transforms/InferStateProperties.cpp [hunk 8]
```diff
@@ -370,7 +380,7 @@ EnableInfo computeEnableInfoFromPattern(OpOperand &output, StateOp stateOp) {
 
 namespace {
 struct InferStatePropertiesPass
-    : public InferStatePropertiesBase<InferStatePropertiesPass> {
+    : public arc::impl::InferStatePropertiesBase<InferStatePropertiesPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/InferStateProperties.cpp:383]: Unrelated changes

File: lib/Dialect/Arc/Transforms/InlineArcs.cpp [hunk 9]
```diff
@@ -6,20 +6,29 @@
...
+} // namespace circt
+
 using namespace circt;
 using namespace arc;
 using mlir::InlinerInterface;
 
 namespace {
 
-struct InlineArcsPass : public InlineArcsBase<InlineArcsPass> {
+struct InlineArcsPass : public arc::impl::InlineArcsBase<InlineArcsPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/InlineArcs.cpp:31]: Unrelated changes

File: lib/Dialect/Arc/Transforms/InlineModules.cpp [hunk 10]
```diff
@@ -6,24 +6,35 @@
...
 using namespace circt;
 using namespace arc;
 using namespace hw;
 using namespace igraph;
 using mlir::InlinerInterface;
 
 namespace {
-struct InlineModulesPass : public InlineModulesBase<InlineModulesPass> {
+struct InlineModulesPass
+    : public arc::impl::InlineModulesBase<InlineModulesPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/InlineModules.cpp:37]: Unrelated changes

File: lib/Dialect/Arc/Transforms/IsolateClocks.cpp [hunk 11]
```diff
@@ -156,7 +165,8 @@ void ClockDomain::computeCrossingValues(SmallVectorImpl<Value> &inputs,
 //===----------------------------------------------------------------------===//
 
 namespace {
-struct IsolateClocksPass : public IsolateClocksBase<IsolateClocksPass> {
+struct IsolateClocksPass
+    : public arc::impl::IsolateClocksBase<IsolateClocksPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/IsolateClocks.cpp:169]: Unrelated changes

File: lib/Dialect/Arc/Transforms/LatencyRetiming.cpp [hunk 12]
```diff
@@ -112,7 +121,8 @@ LatencyRetimingPattern::matchAndRewrite(StateOp op,
 //===----------------------------------------------------------------------===//
 
 namespace {
-struct LatencyRetimingPass : LatencyRetimingBase<LatencyRetimingPass> {
+struct LatencyRetimingPass
+    : arc::impl::LatencyRetimingBase<LatencyRetimingPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/LatencyRetiming.cpp:125]: Unrelated changes

File: lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp [hunk 13]
```diff
@@ -534,7 +543,7 @@ moveMemoryWritesAfterLastRead(Region &region, const DenseSet<Value> &memories,
 
 namespace {
 struct LegalizeStateUpdatePass
-    : public LegalizeStateUpdateBase<LegalizeStateUpdatePass> {
+    : public arc::impl::LegalizeStateUpdateBase<LegalizeStateUpdatePass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp:546]: Unrelated changes

File: lib/Dialect/Arc/Transforms/LowerClocksToFuncs.cpp [hunk 14]
```diff
@@ -24,7 +33,7 @@ using mlir::OpTrait::ConstantLike;
 
 namespace {
 struct LowerClocksToFuncsPass
-    : public LowerClocksToFuncsBase<LowerClocksToFuncsPass> {
+    : public arc::impl::LowerClocksToFuncsBase<LowerClocksToFuncsPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/LowerClocksToFuncs.cpp:36]: Unrelated changes

File: lib/Dialect/Arc/Transforms/LowerLUT.cpp [hunk 15]
```diff
@@ -341,7 +351,7 @@ struct LutToArray : LutLoweringPattern {
 namespace {
 
 /// Lower LutOp operations to comb and hw operations.
-struct LowerLUTPass : public LowerLUTBase<LowerLUTPass> {
+struct LowerLUTPass : public arc::impl::LowerLUTBase<LowerLUTPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/LowerLUT.cpp:354]: Unrelated changes

File: lib/Dialect/Arc/Transforms/LowerState.cpp [hunk 16]
```diff
@@ -683,7 +692,7 @@ LogicalResult ModuleLowering::cleanup() {
 //===----------------------------------------------------------------------===//
 
 namespace {
-struct LowerStatePass : public LowerStateBase<LowerStatePass> {
+struct LowerStatePass : public arc::impl::LowerStateBase<LowerStatePass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/LowerState.cpp:695]: Unrelated changes

File: lib/Dialect/Arc/Transforms/LowerVectorizations.cpp [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+    VectorType type = VectorType::get(SmallVector<int64_t>(1, range.size()),
+                                      range.front().getType());
+    if (llvm::all_equal(range)) {
+      vectors.push_back(
+          builder.create<vector::BroadcastOp>(type, range.front())
+              ->getResults());
+      continue;
+    }
+
+    // Otherwise to a gather.
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/LowerVectorizations.cpp:127]:
```suggestion
    // Otherwise do a gather.
```

File: lib/Dialect/Arc/Transforms/LowerVectorizations.cpp [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+  if (op.isBoundaryVectorized())
+    return op;
+
+  // If the body is already vectorized, we must use the same vectorization
+  // technique. Otherwise, we would produce invalid IR.
+  if (op.isBodyVectorized()) {
+    if (isa<VectorType>(op.getBody().front().getArgumentTypes().front()))
+      return lowerBoundaryVector(op);
+    return lowerBoundaryScalar(op);
+  }
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/LowerVectorizations.cpp:174]: This is very elegant :sunglasses:

File: lib/Dialect/Arc/Transforms/MakeTables.cpp [hunk 19]
```diff
@@ -22,7 +32,7 @@ namespace {
 static constexpr int tableMinOpCount = 20;
 static constexpr int tableMaxSize = 32768; // bits
 
-struct MakeTablesPass : public MakeTablesBase<MakeTablesPass> {
+struct MakeTablesPass : public arc::impl::MakeTablesBase<MakeTablesPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/MakeTables.cpp:35]: Unrelated changes

File: lib/Dialect/Arc/Transforms/MuxToControlFlow.cpp [hunk 20]
```diff
@@ -32,7 +41,7 @@ namespace {
 
 /// Convert muxes to if-statements.
 struct MuxToControlFlowPass
-    : public MuxToControlFlowBase<MuxToControlFlowPass> {
+    : public arc::impl::MuxToControlFlowBase<MuxToControlFlowPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/MuxToControlFlow.cpp:44]: Unrelated changes

File: lib/Dialect/Arc/Transforms/PrintStateInfo.cpp [hunk 21]
```diff
@@ -37,13 +46,14 @@ struct ModelInfo {
...
-struct PrintStateInfoPass : public PrintStateInfoBase<PrintStateInfoPass> {
+struct PrintStateInfoPass
+    : public arc::impl::PrintStateInfoBase<PrintStateInfoPass> {
   void runOnOperation() override;
   LogicalResult runOnOperation(llvm::raw_ostream &outputStream);
   LogicalResult collectStates(Value storage, unsigned offset,
                               std::vector<StateInfo> &stateInfos);
 
-  using PrintStateInfoBase::stateFile;
+  using arc::impl::PrintStateInfoBase<PrintStateInfoPass>::stateFile;
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/PrintStateInfo.cpp:56]: Unrelated changes

File: lib/Dialect/Arc/Transforms/SimplifyVariadicOps.cpp [hunk 22]
```diff
@@ -6,19 +6,29 @@
...
+
 using namespace mlir;
 using namespace circt;
 using namespace arc;
 using namespace hw;
 
 namespace {
 struct SimplifyVariadicOpsPass
-    : public SimplifyVariadicOpsBase<SimplifyVariadicOpsPass> {
+    : public arc::impl::SimplifyVariadicOpsBase<SimplifyVariadicOpsPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/SimplifyVariadicOps.cpp:31]: Unrelated changes

File: lib/Dialect/Arc/Transforms/SplitLoops.cpp [hunk 23]
```diff
@@ -172,7 +181,7 @@ Split &Splitter::getSplit(const APInt &color) {
 //===----------------------------------------------------------------------===//
 
 namespace {
-struct SplitLoopsPass : public SplitLoopsBase<SplitLoopsPass> {
+struct SplitLoopsPass : public arc::impl::SplitLoopsBase<SplitLoopsPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/SplitLoops.cpp:184]: Unrelated changes

File: lib/Dialect/Arc/Transforms/StripSV.cpp [hunk 24]
```diff
@@ -6,21 +6,30 @@
...
+#include "circt/Dialect/Arc/ArcPasses.h.inc"
+} // namespace arc
+} // namespace circt
+
 using namespace circt;
 using namespace arc;
 
 namespace {
-struct StripSVPass : public StripSVBase<StripSVPass> {
+struct StripSVPass : public arc::impl::StripSVBase<StripSVPass> {
```
Comment (fabianschuiki) [inline lib/Dialect/Arc/Transforms/StripSV.cpp:32]: Unrelated changes

**Review by fabianschuiki — APPROVED** [review]
This is really cool! LGTM :sunglasses: Almost makes me think that `comb` ops should support the `vector` type.

For some reason GH still shows a bunch of unrelated `PassDetails.h` changes that were already landed. Weird.

**Comment by maerhart** [comment]
> For some reason GH still shows a bunch of unrelated PassDetails.h changes that were already landed. Weird.

This is really strange. I opened this as a stacked PR, but it seems like the automatic base branch change of GH didn't work properly eventhough I explicitly did a rebase merge on the other PR instead of a squash. I'll make sure that only the changes in the top commit will land with this PR.
